Index: channels/chan_sip.c
===================================================================
--- channels/chan_sip.c	(revision 402886)
+++ channels/chan_sip.c	(working copy)
@@ -21113,6 +21113,8 @@
 		}
 		break;
 
+	case 414: /* Bad request URI */
+	case 493: /* Undecipherable */
 	case 404: /* Not found */
 		xmitres = transmit_request(p, SIP_ACK, seqno, XMIT_UNRELIABLE, FALSE);
 		if (p->owner && !req->ignore) {
@@ -22010,6 +22012,15 @@
 			else
 				ast_log(LOG_WARNING, "Host '%s' does not implement '%s'\n", ast_sockaddr_stringify(&p->sa), msg);
 			break;
+		case 400: /* Bad Request */
+		case 414: /* Request URI too long */
+		case 493: /* Undecipherable */
+			if (sipmethod == SIP_INVITE) {
+				handle_response_invite(p, resp, rest, req, seqno);
+			} else {
+				ast_log(LOG_WARNING, "Bad request to '%s': '%s'\n", ast_sockaddr_stringify(&p->sa), msg);
+			}
+			break;
 		default:
 			if ((resp >= 300) && (resp < 700)) {
 				/* Fatal response */
