Index: main/channel.c
===================================================================
--- main/channel.c	(revision 402889)
+++ main/channel.c	(working copy)
@@ -7559,8 +7559,10 @@
 	manager_bridge_event(1, 1, c0, c1);
 
 	/* Before we enter in and bridge these two together tell them both the source of audio has changed */
+#ifdef SRCUPDATE_ALWAYS
 	ast_indicate(c0, AST_CONTROL_SRCUPDATE);
 	ast_indicate(c1, AST_CONTROL_SRCUPDATE);
+#endif
 
 	for (/* ever */;;) {
 		struct timeval now = { 0, };
@@ -7733,8 +7735,10 @@
 	ast_clear_flag(c1, AST_FLAG_END_DTMF_ONLY);
 
 	/* Now that we have broken the bridge the source will change yet again */
+#ifdef SRCUPDATE_ALWAYS
 	ast_indicate(c0, AST_CONTROL_SRCUPDATE);
 	ast_indicate(c1, AST_CONTROL_SRCUPDATE);
+#endif
 
 	ast_channel_lock_both(c0, c1);
 	c0->_bridge = NULL;
Index: channels/chan_sip.c
===================================================================
--- channels/chan_sip.c	(revision 402903)
+++ channels/chan_sip.c	(working copy)
@@ -7300,6 +7300,7 @@
 		}
 		break;
 	case AST_CONTROL_SRCUPDATE:
+		ast_debug(3, "Got control frame AST_CONTROL_SRCUPDATE on %s\n", p->callid);
 		ast_rtp_instance_update_source(p->rtp);
 		break;
 	case AST_CONTROL_SRCCHANGE:
