Index: channels/chan_sip.c
===================================================================
--- channels/chan_sip.c	(revision 417744)
+++ channels/chan_sip.c	(working copy)
@@ -9603,7 +9603,7 @@
 				if (audio) {
 					if (process_sdp_a_sendonly(value, &sendonly)) {
 						processed = TRUE;
-					} else if (!processed_crypto && process_crypto(p, p->rtp, &p->srtp, value)) {
+					} else if (secure_audio && !processed_crypto && process_crypto(p, p->rtp, &p->srtp, value)) {
 						processed_crypto = TRUE;
 						processed = TRUE;
 					} else if (process_sdp_a_audio(value, p, &newaudiortp, &last_rtpmap_codec)) {
